### PR TITLE
Run tests in CI

### PR DESCRIPTION
**Pull request by Julian**
Some GitHub actioning, a tox setup, some minor packaging things, a LICENSE, and pre-commit, which is highly useful (though yeah let me know if you don't want/like any of the above and obviously can drop it, this is just essentially a "standard" setup I have generated for anything I've got).

This will undoubtedly conflict with #19 but should be somewhat easy to rebase (happy to do that too of course).

### CODE REVIEW

**Comment by Julian** [comment]
(And this will undoubtedly fail, even when I get the kinks out, given #18, but yeah will see...)

**Comment by Julian** [comment]
OK these are failing for real reasons I think now. (Though yeah some likely affected by #19) -- the [first one being subclassing something in trio (which generally doesn't like things subclassing it)](https://github.com/Julian/lean-client-python/runs/1350304092?check_suite_focus=true#step:10:72), which fails during import, so tests haven't started running yet even.

**Comment by Julian** [comment]
Uh, whoops, I didn't notice #12 either, which this would close, but hopefully @jasonrute you'll forgive me for preempting your attempt to understand how said stuff works better :D

**Comment by jasonrute** [comment]
Julian, I for one would be more than happy for you to take over CI (#12).  I did some initial research on it, and then got caught up on other things.

**Comment by jasonrute** [comment]
@Julian If this is going to auto format, I'd prefer to put it in after my massive refactor.

**Comment by jasonrute** [comment]
But other than that, my only question is if it also runs mypy.  I've found mypy very helpful for this project with all the type stuff and it would be good for it to be run in CI.

**Comment by Julian** [comment]
> But other than that, my only question is if it also runs mypy. I've found mypy very helpful for this project with all the type stuff and it would be good for it to be run in CI.

I didn't add it, but I'm happy to.

> @Julian If this is going to auto format, I'd prefer to put it in after my massive refactor.

Want me to split that piece off and just do the CI first for a first PR? Or just wait on all of it? Either way seems fine with me.

**Comment by jasonrute** [comment]
> Want me to split that piece off and just do the CI first for a first PR? Or just wait on all of it? Either way seems fine with me.

If we can get the other one in soon, let's wait.  If not, then let's break it up and push the CI in without all the formatting changes.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      with:
+        python-version: ${{ matrix.python-version.name }}
+    - name: Ensure we have new enough versions to respect python_version
+      run: python -m pip install -U pip setuptools
+    - name: Install dependencies
+      run: brew install gmp coreutils
+      if: runner.os == 'macOS'
+    - name: Install elan
+      run: |
+          curl https://raw.githubusercontent.com/Kha/elan/master/elan-init.sh -sSf | sh -s -- --default-toolchain leanprover-community/lean:3.23.0 -y
```
Comment (bryangingechen) [inline .github/workflows/ci.yml:60]:
```suggestion
          curl https://raw.githubusercontent.com/leanprover/elan/master/elan-init.sh -sSf | sh -s -- --default-toolchain leanprover-community/lean:3.23.0 -y
```

Comment (bryangingechen) [reply .github/workflows/ci.yml:60]: Oh, I just noticed: are you sure you want to hardcore a specific Lean version here?

Comment (Julian) [reply .github/workflows/ci.yml:60]: Almost certainly not, probably we want a matrix build with whatever lean versions we claim to support, but I think I copied this from elsewhere, probably mathlib-tools, which doesn't yet do that, it hardcodes one version.

(Which is probably a poor reason, I can just start doing it here I guess, so yeah will give that a shot).

I do though owe Jason a review on the commands API PR which I've procrastinated.. will probably try to do that before twiddling more here.

Thanks for the review!

**Pull request closed without merging**